File: src/main/java/com/kishlaly/ta/analyze/tasks/blocks/one/Long_ScreenOne_EMA_LastCrossesOrAbove.java
package com.kishlaly.ta.analyze.tasks.blocks.one;

import com.kishlaly.ta.model.BlockResultJava;
import com.kishlaly.ta.model.QuoteJava;
import com.kishlaly.ta.model.SymbolDataJava;
import com.kishlaly.ta.model.indicators.EMAJava;
import com.kishlaly.ta.model.indicators.IndicatorJava;
import com.kishlaly.ta.utils.CollectionUtilsJava;
import com.kishlaly.ta.utils.LogJava;
import com.kishlaly.ta.utils.QuotesJava;

import java.util.List;

import static com.kishlaly.ta.model.BlockResultCodeJava.LAST_QUOTE_NOT_CROSSING_OR_NOT_ABOVE_EMA_SCREEN_1;
import static com.kishlaly.ta.model.BlockResultCodeJava.OK;

/**
 * the last bar crosses the EMA26 or higher
 */
public class Long_ScreenOne_EMA_LastCrossesOrAbove implements ScreenOneBlockJava {
    @Override
    public BlockResultJava check(SymbolDataJava screen) {
        QuoteJava lastQuote = screen.getLastQuote();
        List<EMAJava> screen_2_EMA26 = (List<EMAJava>) screen.indicators.get(IndicatorJava.EMA26);
        EMAJava lastEMA26 = CollectionUtilsJava.getFromEnd(screen_2_EMA26, 1);
        boolean lastBarCrossing = QuotesJava.isQuoteCrossedEMA(lastQuote, lastEMA26.getValue());
        boolean lastBarAbove = QuotesJava.isQuoteAboveEMA(lastQuote, lastEMA26.getValue());

        if (!lastBarCrossing || !lastBarAbove) {
            LogJava.recordCode(LAST_QUOTE_NOT_CROSSING_OR_NOT_ABOVE_EMA_SCREEN_1, screen);
            LogJava.addDebugLine("The last bar does not cross or is not above the EMA on the long term screen");
            return new BlockResultJava(lastQuote, LAST_QUOTE_NOT_CROSSING_OR_NOT_ABOVE_EMA_SCREEN_1);
        }
        return new BlockResultJava(lastQuote, OK);
    }
}

File: src/main/kotlin/com/kishlaly/ta/analyze/tasks/blocks/one/Long_ScreenOne_EMA_LastCrossesOrAbove.kt
package com.kishlaly.ta.analyze.tasks.blocks.one

import com.kishlaly.ta.model.*
import com.kishlaly.ta.model.indicators.EMA
import com.kishlaly.ta.model.indicators.Indicator
import com.kishlaly.ta.utils.Log

/**
 * the last bar crosses the EMA26 or higher
 */
class Long_ScreenOne_EMA_LastCrossesOrAbove : ScreenOneBlock {
    override fun check(screen: SymbolData): BlockResult {
        val screen_1_EMA26 = screen.indicator(Indicator.EMA26) as List<EMA>
        val lastBarCrossing = screen.lastQuote.crosses(screen_1_EMA26.last())
        val lastBarAbove = screen.lastQuote above screen_1_EMA26.last()

        if (!lastBarCrossing || !lastBarAbove) {
            Log.recordCode(BlockResultCode.LAST_QUOTE_NOT_CROSSING_OR_NOT_ABOVE_EMA_SCREEN_1, screen)
            Log.addDebugLine("The last bar does not cross or is not above the EMA on the long term screen")
            return BlockResult(screen.lastQuote, BlockResultCode.LAST_QUOTE_NOT_CROSSING_OR_NOT_ABOVE_EMA_SCREEN_1)
        }

        return BlockResult(screen.lastQuote, BlockResultCode.OK)
    }
}